File: src/java/pkg/Interpreter.java
package pkg;

//import static pkg.Util.isDouble;

import java.util.List;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.Map;
import java.util.HashMap;

/**
 * A simple interpreter that can calculate arbitrarily nested mathematical
 * expressions.
 * It supports:
 * - operators +, -, *, /, %.
 * - variable assignments, e.g. x = 1 or x = 1 + 1
 * - variable value retrieval
 * - operation nesting
 */
public class Interpreter {

    private final Map<String, Double> variables = new HashMap<>();
    private static final String OPERATORS = "+-*/%";

    /**
     * Tokenizes and interprets the specified string, supporting variable assignment
     * and arbitrarily nested mathematical expressions.
     * Throws a SyntaxException for invalid input.
     */
    public Double interpret(final String string) {
        List<String> tokens = Tokenizer.tokenize(string);
        if (!isBalancedParentheses(tokens))
            throw new SyntaxException("Unbalanced parentheses.");
        Double result = interpret(tokens);
        return result;
    }

    /** @see Interpreter#interpret(String) */
    public Double interpret(final List<String> tokens) {
        if (tokens.isEmpty()) {
            return null;
        }

        // check for assignment
        if (tokens.size() > 2 && tokens.get(1).equals("=")) {
            String variable = tokens.remove(0);
            tokens.remove(0); // assignment operator
            double value = interpret(tokens);
            variables.put(variable, value);
            return value;
        }

        Expression root = new Expression();
        Expression current = root;
        while (tokens.size() > 0) {
            if (current.isSatisfied()) {
                if (current.isLowPrecedence()) {
                    current = current.createSubExpression();
                } else {
                    root = new Expression(current);
                    current = root;
                }
            }

            String token = tokens.remove(0);
            if (token.equals(")")) {
                return root.getValue();
            }
            if (token.equals("(")) {
                current.pushValue(interpret(tokens));
            } else if (Util.isDouble(token)) {
                current.pushValue(Double.parseDouble(token));
            } else if (OPERATORS.contains(token)) {
                current.setOperator(token);
            } else {
                try {
                    current.pushValue(variables.get(token));
                } catch (Exception exc) {
                    throw new RuntimeException("Undefined variable " + token);
                }
            }
        }

        return root.getValue();
    }

    private static boolean isBalancedParentheses(final List<String> tokens) {
        int scopes = 0;
        for (var token : tokens) {
            if (token == "(")
                scopes++;
            else if (token == ")")
                scopes--;
        }
        return scopes == 0;
    }
}

File: src/c/src/interpreter.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#include "interpreter.h"

int MAX_INPUT_LENGTH = 1000;

/* Reads a null-terminated line into the specified buffer. If
a newline char isn't found before the specified maxlen, returns -1,
else returns 0.
*/
int getline(char *buffer, int maxlen)
{
    int i = 0;
    char c;
    while (i < maxlen)
    {
        c = getchar();
        buffer[i] = c;
        if (c == '\n')
        {
            buffer[i] = '\0';
            return 0;
        }
        i++;
    }
    return -1;
}

/* Prints the interpreter help doc */
void printHelp()
{
    printf("A simple interpreter that can calculate arbitrarily nested mathematical ");
    printf("expressions.\nIt supports:\n");
    printf("- operators +, -, *, /, %%.\n");
    printf("- variable assignments, e.g. x = 1 or x = 1 + 1 (max 100 variables)\n");
    printf("- variable value retrieval\n");
    printf("- operation nesting\n");
    printf("Note that it is integer-based and does not understand floating point numbers");
    printf(" or float division.\n");
}

/* Handles StatusCodes defined in interpreter.h */
int handleErrorStatus(const int status, const int *result)
{
    switch (status)
    {
    case 0:
        printf("%i\n", *result);
        break;
    case 1:
        break;
    case 2:
        printf("Invalid operator\n");
        break;
    case 3:
        printf("Undefined variable\n");
        break;
    case 4:
        printf("Invalid input\n");
        break;
    case 5:
        printf("Division by zero\n");
        break;
    default:
        printf("Error occured\n");
        break;
    }
}

/* A simple REPL interpreter for integer operations. */
int main()
{
    char *buffer = malloc(sizeof(char) * MAX_INPUT_LENGTH);
    size_t size;
    int status = 0;
    int result[] = {0};
    initInterpreter();
    printf("Simple REPL shell. Type \"help\" for help, or \"exit\" to exit the shell.\n");
    while (true)
    {
        printf(">> ");
        status = getline(buffer, MAX_INPUT_LENGTH);
        if (status == -1)
            continue;
        if (strncmp(buffer, "exit", MAX_INPUT_LENGTH) == 0)
            break;
        if (strncmp(buffer, "help", MAX_INPUT_LENGTH) == 0)
        {
            printHelp();
            continue;
        }
        status = evaluate(buffer, result);
        handleErrorStatus(status, result);
    }
    closeInterpreter();
    free(buffer);
    return 0;
}
